source/validate_atomics.cpp

```diff
@@ -0,0 +1,260 @@
+      }
+
+      if (!_.IsFloatScalarType(data_type) &&
+          !_.IsIntScalarType(data_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to be a pointer to int or float "
+               << "scalar type";
+      }
+
+      const uint32_t scope_type = _.GetOperandTypeId(inst, 1);
+      if (!_.IsIntScalarType(scope_type) || _.GetBitWidth(scope_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Scope to be 32-bit int";
+      }
+
+      const uint32_t memory_semantics_type = _.GetOperandTypeId(inst, 2);
+      if (!_.IsIntScalarType(memory_semantics_type) ||
+          _.GetBitWidth(memory_semantics_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Memory Semantics to be 32-bit int";
+      }
+
+      if (opcode == SpvOpAtomicStore) {
+        const uint32_t value_type = _.GetOperandTypeId(inst, 3);
+        if (value_type != data_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << spvOpcodeString(opcode)
+              << ": expected Value type and the type pointed to by Pointer to"
+              << " be the same";
+        }
+      }
+
+      break;
+    }
+
+    case SpvOpAtomicFlagTestAndSet: {
+      if (!_.IsBoolScalarType(result_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << spvOpcodeString(opcode)
+            << ": expected Result Type to be bool scalar type";
+      }
+
+      const uint32_t pointer_type = _.GetOperandTypeId(inst, 2);
+
+      uint32_t data_type = 0;
+      uint32_t storage_class = 0;
+      if (!_.GetPointerTypeInfo(pointer_type, &data_type, &storage_class)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to be of type OpTypePointer";
+      }
+
+      if (!_.IsIntScalarType(data_type) || _.GetBitWidth(data_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Pointer to point to a value of 32-bit int type";
+      }
+
+      const uint32_t scope_type = _.GetOperandTypeId(inst, 3);
+      if (!_.IsIntScalarType(scope_type) || _.GetBitWidth(scope_type) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Scope to be 32-bit int";
+      }
+
+      const uint32_t memory_semantics_type1 = _.GetOperandTypeId(inst, 4);
+      if (!_.IsIntScalarType(memory_semantics_type1) ||
+          _.GetBitWidth(memory_semantics_type1) != 32) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+               << spvOpcodeString(opcode)
+               << ": expected Memory Semantics to be 32-bit int";
+      }
+
+      break;
+    }
+
+    case SpvOpAtomicFlagClear: {
```

Comment:
I feel this case can be merged with the case for `SpvOpAtomicStore` to share some logic?

Reply:
Done.